AdyenCardScanner/Sources/CardImageParser.swift

```diff
@@ -0,0 +1,233 @@
+        colorControlsFilter?.setValue(0.0, forKey: kCIInputSaturationKey) // No color
+        colorControlsFilter?.setValue(1.0, forKey: kCIInputBrightnessKey) // Full brightness
+        colorControlsFilter?.setValue(1.0, forKey: kCIInputContrastKey) // Full contrast
+        return colorControlsFilter?.outputImage
+    }
+
+    private func applyThreshold(image: CIImage, threshold: CGFloat) -> CIImage? {
+        let colorMatrixFilter = CIFilter(name: "CIColorMatrix")
+        colorMatrixFilter?.setValue(image, forKey: kCIInputImageKey)
+
+        // Set the threshold value
+        let thresholdVector = CIVector(x: threshold, y: threshold, z: threshold, w: 1)
+        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputRVector")
+        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputGVector")
+        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputBVector")
+
+        // Keeping alpha channel as is
+        colorMatrixFilter?.setValue(CIVector(x: 0, y: 0, z: 0, w: 1), forKey: "inputAVector")
+
+        // Biasing the output so that it’s either black or white
+        let biasVector = CIVector(x: -threshold, y: -threshold, z: -threshold, w: 0)
+        colorMatrixFilter?.setValue(biasVector, forKey: "inputBiasVector")
+
+        return colorMatrixFilter?.outputImage
+    }
+
+    // ==================================================
+
+    private func extractCardNumber(from textObservations: [VNRecognizedTextObservation]) -> String? {
+        for observation in textObservations {
+            let candidates = observation.topCandidates(Constants.topCandidates)
+            guard let topCandidate = candidates.first,
+                  topCandidate.confidence > Constants.cardNumberConfidence else { continue }
+
+            debugPrint("candidates >> \(topCandidate.string)")
+
+            let sanitizedCardNumber = topCandidate.string.replacingOccurrences(of: " ", with: "")
+
+            guard sanitizedCardNumber.isOnlyNumbers else { continue }
+            guard sanitizedCardNumber.count >= 13, sanitizedCardNumber.count <= 19 else { continue }
+            guard isValidLuhn(sanitizedCardNumber) else { continue }
+
+            debugPrint("VALID CARD: \(sanitizedCardNumber)")
+
+            return sanitizedCardNumber
+        }
+
+        return nil
+    }
+
+    private func extractExpireDate(from textObservations: [VNRecognizedTextObservation]) -> Date? {
+        for observation in textObservations {
+            let candidates = observation.topCandidates(Constants.topCandidates)
+            guard let topCandidate = candidates.first,
+                  topCandidate.confidence > Constants.expireDateConfidence else { continue }
+
+            let expireDateMatch = extractMatch(from: topCandidate.string, using: Constants.expireDateRegex)
+            guard let expireDateMatch else { continue }
+
+            debugPrint("VALID DATE: \(expireDateMatch)")
```

Comment:
Better use `AdyenLogging`

Reply:
This debug prints are gonna dissappear. I want to prevent any dependency with other modules.